### PR TITLE
Extra data filter

### PR DESCRIPTION
**Pull request by federicoB**
Added code in Annotation Search view to filter annotation also by exact key-value in extra data.

Completed with test.

### CODE REVIEW

**Comment by rlskoeser** [comment]
@federicoB I apologize!  I missed this somehow.

File: annotator_store/views.py [hunk 1]
```diff
@@ -236,6 +236,17 @@ def get(self, request):
                 )
                 # NOTE: contains search on extra data jsonfield is
                 # probably not a great idea...
+            # non-standard field search on extra data
+            # ignore limit and offset field because they must be used afterward
+            elif field not in ['limit','offset']:
+                # evaluate queryset for extra_data JSONField parsing
+                annotations = notes.all()
```
Comment (rlskoeser) [inline annotator_store/views.py]: Am I following this correctly that you're pulling all the annotations down and filtering them locally?  That won't scale.   Could you look at a solution like some of the suggestions here? https://stackoverflow.com/questions/10445176/how-to-write-a-query-to-get-find-value-in-a-json-field-in-django

File: annotator_store/views.py [hunk 2]
```diff
@@ -236,6 +236,17 @@ def get(self, request):
                 )
                 # NOTE: contains search on extra data jsonfield is
                 # probably not a great idea...
+            # non-standard field search on extra data
```
Comment (rlskoeser) [inline annotator_store/views.py:243]: Need to add documentation to the view method about this type of search being supported

**Comment by rlskoeser** [comment]
@federicoB could you elaborate a bit for me, what is the value of this feature?  What is your use case?

File: annotator_store/views.py (hunk 1 above)
Comment (federicoB) [reply annotator_store/views.py]: Yes, that won't scale. The __contains solution is better. I didn't think about using it with json format. In a way it's searches exactly that value, not what usually contains do. I'm going to implement it!

**Comment by federicoB** [comment]
If we want to keep the extra_data as a json blob useful only to the frontend this feature doesn't have much value.
I created this feature for filtering down annotation based on web pages in which they were created.
I just realized that i could use a the uri field for that.
Also if anyone ever wanted to add fields to his annotations and wanted to filter for these, he could extend the model and put the new fields in the common_fields.

So, if you like the solution of __contains and allow to filter the annotations by extra data value, I can change this pull request otherwise you can close it and end it here.

**Comment by rlskoeser** [comment]
@federicoB yes, use the uri field for filtering annotations by associated page (that's how annotator.js loads them).  

The idea is for people to extend the annotation model if they need custom fields for more than just storage and retrieval.   This probably means we should also make the `AnnotationSearch` view easier to extend so you can customize just the search logic (eventually!).

However, I like the idea of the `__contains` filter with json formatting, if you can get it to work. (The conversation on StackOverflow made it sound like it could be problematic, but seems like it should work well enough to be useful.)

**Review by rlskoeser — CHANGES_REQUESTED** [review]
Reimplement to use `__contains` filter with json-encoded values and document as a search option.

**Comment by rlskoeser** [comment]
@federicoB just documenting what we discussed so the PR is marked as changes requested.

**Comment by federicoB** [comment]
Lots of nice lookups methods for JSONFields [are available only with PostgreSQL](https://docs.djangoproject.com/en/2.0/ref/contrib/postgres/fields/#querying-jsonfield).

I tried the suggested solution from StackOverflow but the lookup fail and the query returns nothing.
I also tried to manually build a string without using json.dump() but no success.

**Comment by federicoB** [comment]
Provided a simple but more performing solution.

Chained __contains lookup for key and value.
This can cause false positives when both key and value are inside a string used only as key or value, or when key and value are present, but in different elements of the dict.

Better than nothing, choose if merge this or wait if someone else find a better solution.

**Comment by rlskoeser** [comment]
I'm not happy with this solution (the contains matches as written are too general), and have kept the PR open thinking I would get around to revising it myself at some point, but it's been almost 2 years so clearly that's not happening. I'm going to go ahead and close it, sorry.

**Pull request closed without merging**